apps/meteor/server/models/raw/LivechatContacts.ts

```diff
@@ -61,11 +110,132 @@ export class LivechatContactsRaw extends BaseRaw<ILivechatContact> implements IL
 		);
 	}
 
+	async findContactMatchingVisitor(visitor: AtLeast<ILivechatVisitor, 'visitorEmails' | 'phone'>): Promise<ILivechatContact | null> {
+		const emails = visitor.visitorEmails?.map(({ address }) => address).filter((email) => Boolean(email)) || [];
+		const phoneNumbers = visitor.phone?.map(({ phoneNumber }) => phoneNumber).filter((phone) => Boolean(phone)) || [];
+
+		if (!emails?.length && !phoneNumbers?.length) {
+			return null;
+		}
+
+		const query = {
+			$and: [
+				{
+					$or: [
+						...emails?.map((email) => ({ 'emails.address': email })),
+						...phoneNumbers?.map((phone) => ({ 'phones.phoneNumber': phone })),
+					],
+				},
+				{
+					$or: [
+						{
+							channels: { $exists: false },
+						},
+						{
+							channels: { $size: 0 },
+						},
+					],
+				},
+			],
+		};
+
+		return this.findOne(query);
+	}
+
+	private makeQueryForVisitor(
+		visitor: ILivechatContactVisitorAssociation,
+		extraFilters?: Filter<Required<ILivechatContact>['channels'][number]>,
+	): Filter<ILivechatContact> {
+		return {
+			channels: {
```

Comment:
we *must* create indexes for queries created by this method.. need to test but I think it could be an index like:
```
{
  'channels.visitor.visitorId': 1,
  'channels.visitor.source.type': 1,
  'channels.visitor.source.id': 1,
}
```

Reply:
It works! I created a compound index as you suggested